### PR TITLE
Fix bugs in PosControllers

### PR DESCRIPTION
**Pull request by ThijsLuttikhuis**
- [ ] The tests are passing

### Summary
Give a short summary of what the changes are and why you made them. If applicable, apply screenshots as well. 

### Related issues
If there are related issues, reference them here.

fixes ...  
fixes ...

### Comments for the reviewer
If there are comments for the reviewer, leave them here

### Checklist for the reviewer
This acts as a guide to help you and/or the reviewer to analyze the code. 

**functionality**  
- Is the code is properly tested?
- Is the code working as expected
- Is the code is understandable and easy to read
- Are proper abstractions made? (no duplicate code, nice use of functions/objects)

**style**  
- Variables/objects/classes have proper names
- Proper use of variables: nu unused vars, no magic numbers
- No possible memory leaks. Mutexes properly used when multithreading? possible pointer issues?
- Are smart pointers used?
- unused variables?
- magic numbers?
- well documented?
- No unused imports?
- No leftover code / unused comments
- Proper indentation
- Proper use of namespaces

### CODE REVIEW

File: roboteam_ai/src/control/ballHandling/BallHandlePosControl.cpp [hunk 1]
```diff
@@ -250,17 +269,17 @@ RobotCommand BallHandlePosControl::goToMovingBall() {
...
     bool robotIsBehindBall = fabs(robotAngleTowardsBallVel) < M_PI_4;
 
-    bool robotCanInterceptBall = robotToProjectionDistance/averageRobotVelocity <
+    bool robotCanInterceptBall = robotToProjectionDistance/averageRobotInterceptVelocity <
             ballToProjectionDistance/ball->vel.length();
 
     if (robotIsBehindBall && robotCanInterceptBall) {
         numTreesTarget = projectionPosition;
-        auto robotCommand = NumTreePosControl::getRobotCommand(robot, numTreesTarget);
+        auto robotCommand = BasicPosControl::getRobotCommand(robot, numTreesTarget);
```
Comment (RobZelluf) [inline roboteam_ai/src/control/ballHandling/BallHandlePosControl.cpp]: Take the posController of the robot

File: roboteam_ai/src/control/ballHandling/BallHandlePosControl.cpp [hunk 2]
```diff
@@ -298,7 +321,18 @@ RobotCommand BallHandlePosControl::goToIdleBall(
     else {
         numTreesTarget = ball->pos + ballToTarget.stretchToLength(maxBallDistance);
     }
-    return NumTreePosControl::getRobotCommand(robot, numTreesTarget);
+    auto robotCommand = NumTreePosControl::getRobotCommand(robot, numTreesTarget);
```
Comment (RobZelluf) [inline roboteam_ai/src/control/ballHandling/BallHandlePosControl.cpp]: Take posController from robot

File: roboteam_ai/src/control/ControlUtils.cpp [hunk 3]
```diff
@@ -346,7 +346,7 @@ const world::World::RobotPtr ControlUtils::getRobotClosestToLine(std::vector<wor
 }
 
     Vector2 ControlUtils::getInterceptPointOnLegalPosition(Vector2 position, Line line, bool canMoveInDefenseArea, bool canMoveOutOfField, double defenseAreamargin, double outOfFieldMargin) {
-        LineSegment shotLine(line.start, line.end + line.end + (line.end - line.start) * 10000);
+        LineSegment shotLine(line.start, line.end + (line.end - line.start).stretchToLength(100.0));
```
Comment (rolfvdhulst) [inline roboteam_ai/src/control/ControlUtils.cpp]: 100 times is not enough in case of short distances.

File: roboteam_ai/src/control/ballHandling/BallHandlePosControl.cpp [hunk 4]
```diff
@@ -201,23 +210,33 @@ RobotCommand BallHandlePosControl::handleBall(const Vector2 &targetBallPos, Trav
...
+    case BACKWARDS:
+        return updatePID(xBallHandlePID, yBallHandlePID,
+                dribbleBackwards->getRobotCommand(robot, targetBallPos, targetAngle));
+
     default:
     case NO_PREFERENCE: {
         // choose based on distance from the ball to the target
         if (ballIsFarFromTarget) {
-            return dribbleForwards->getRobotCommand(robot, targetBallPos, targetAngle);
+            return updatePID(xBallHandlePID, yBallHandlePID,
```
Comment (rolfvdhulst) [inline roboteam_ai/src/control/ballHandling/BallHandlePosControl.cpp]: The usage of this updatePID function is really vague. You're updating a PID but also passing a robotcommand?

File: roboteam_ai/src/control/ballHandling/BallHandlePosControl.cpp [hunk 5]
```diff
@@ -236,9 +255,9 @@ RobotCommand BallHandlePosControl::goToBall(const Vector2 &targetBallPos, Travel
     bool ballIsMoving = ball->vel.length2() > minVelForMovingball*minVelForMovingball;
 
     if (ballIsMoving) {
-        return goToMovingBall();
+        return updatePID(xGoToBallPID, yGoToBallPID, goToMovingBall());
     }
-    return goToIdleBall(targetBallPos, travelStrategy, ballIsFarFromTarget);
+    return updatePID(xGoToBallPID, yGoToBallPID, goToIdleBall(targetBallPos, travelStrategy, ballIsFarFromTarget));
```
Comment (rolfvdhulst) [inline roboteam_ai/src/control/ballHandling/BallHandlePosControl.cpp]: how would the travelstrategy ever matter for updating the PID?

File: roboteam_ai/src/control/ballHandling/BallHandlePosControl.cpp [hunk 6]
```diff
@@ -298,7 +321,18 @@ RobotCommand BallHandlePosControl::goToIdleBall(
...
     }
-    return NumTreePosControl::getRobotCommand(robot, numTreesTarget);
+    auto robotCommand = NumTreePosControl::getRobotCommand(robot, numTreesTarget);
+    if ((ball->pos - robot->pos).length() < maxBallDistance*1.5) {
+        robotCommand.angle = (ball->pos - robot->pos).toAngle();
+    }
+    return robotCommand;
+}
+
+RobotCommand BallHandlePosControl::updatePID(PID &xpid, PID &ypid, const RobotCommand &robotCommand) {
```
Comment (rolfvdhulst) [inline roboteam_ai/src/control/ballHandling/BallHandlePosControl.cpp]: This is not updating the PID values as the name implies but rather executing it.

File: roboteam_ai/src/control/shotControllers/ShotController.cpp [hunk 7]
```diff
@@ -69,7 +67,7 @@ RobotCommand ShotController::getRobotCommand(world::Robot robot, const Vector2 &
     }
     else {
         isShooting = false;
-        shotData = goToPlaceBehindBall(robot, lineToDriveOver.first + ball->vel*0.2, lineToDriveOver, currentDesiredGeneva);
+        shotData = goToPlaceBehindBall(robot, lineToDriveOver.first, lineToDriveOver, currentDesiredGeneva);
```
Comment (rolfvdhulst) [inline roboteam_ai/src/control/shotControllers/ShotController.cpp]: Why was this removed? Testing? Seems like the robot will just drive against the ball now when it is moving away from the robot which is not behavior you really want

File: roboteam_ai/src/skills/Receive.cpp [hunk 8]
```diff
@@ -46,7 +47,7 @@ Receive::Status Receive::onUpdate() {
         }
     }
 
-    command.w = (ball->pos - robot->pos).toAngle().getAngle();
+    //command.w = (ball->pos - robot->pos).toAngle().getAngle();
```
Comment (rolfvdhulst) [inline roboteam_ai/src/skills/Receive.cpp]: remove commented code.

File: roboteam_ai/src/world/Ball.cpp [hunk 9]
```diff
@@ -168,30 +170,63 @@ void Ball::updateBallPosition(const Ball &oldBall, const WorldData &worldData) {
...
 }
 
+void Ball::filterBallVelocity(Ball &oldBall, const WorldData &worldData) {
+
+    auto &ball = worldData.ball;
+    double velocityDifference = (ball->vel - oldBall.vel).length();
+
+    double b = 8.0;
+    double c = 0.8;
+    double a = velocityDifference > b ? c : velocityDifference*c/b;
```
Comment (rolfvdhulst) [inline roboteam_ai/src/world/Ball.cpp]: variable naming, please.

File: roboteam_ai/src/world/Ball.cpp [hunk 10]
```diff
@@ -168,30 +170,63 @@ void Ball::updateBallPosition(const Ball &oldBall, const WorldData &worldData) {
...
+//        }
+//    }
+
 }
 
 const Vector2 &Ball::getBallStillPosition() const {
     return ballStillPosition;
 }
 
+void Ball::filterBallVelocity(Ball &oldBall, const WorldData &worldData) {
```
Comment (rolfvdhulst) [inline roboteam_ai/src/world/Ball.cpp]: What does this do and why do we need a posthoc filter? This is what Kalman is for.

File: roboteam_ai/src/control/ballHandling/BallHandlePosControl.cpp (hunk 1 above)
Comment (ThijsLuttikhuis) [reply roboteam_ai/src/control/ballHandling/BallHandlePosControl.cpp]: this is indirectly the poscontroller of the robot.
I'm calling the overwritten function of this

File: roboteam_ai/src/control/ControlUtils.cpp (hunk 3 above)
Comment (ThijsLuttikhuis) [reply roboteam_ai/src/control/ControlUtils.cpp]: read: stretchToLength. the field is smaller than 100m so this always works.

File: roboteam_ai/src/control/ballHandling/BallHandlePosControl.cpp (hunk 5 above)
Comment (ThijsLuttikhuis) [reply roboteam_ai/src/control/ballHandling/BallHandlePosControl.cpp]: I should put this on two lines.  'goToIdleBall' is a different function call

File: roboteam_ai/src/control/shotControllers/ShotController.cpp (hunk 7 above)
Comment (ThijsLuttikhuis) [reply roboteam_ai/src/control/shotControllers/ShotController.cpp]: this uses ballhandle now, which automatically handles everything with the ball for you.
this means that you do not have to take anything into account anymore, so also not the ball velocity.

File: roboteam_ai/src/skills/Receive.cpp (hunk 8 above)
Comment (ThijsLuttikhuis) [reply roboteam_ai/src/skills/Receive.cpp]: agreed

File: roboteam_ai/src/world/Ball.cpp (hunk 10 above)
Comment (ThijsLuttikhuis) [reply roboteam_ai/src/world/Ball.cpp]: kalman is very noisy for ball velocity, and that is behaviour which we really do not want for calculations using the ball velocity.

Comment (ThijsLuttikhuis) [reply roboteam_ai/src/world/Ball.cpp]: so I put a moving average over it to make it stable.

File: roboteam_ai/src/control/ballHandling/BallHandlePosControl.cpp (hunk 1 above)
Comment (ThijsLuttikhuis) [reply roboteam_ai/src/control/ballHandling/BallHandlePosControl.cpp]: that is so that we have the same variables for 'avoid defense area' and 'avoid ball' etc etc

File: roboteam_ai/src/control/ballHandling/BallHandlePosControl.cpp (hunk 4 above)
Comment (ThijsLuttikhuis) [reply roboteam_ai/src/control/ballHandling/BallHandlePosControl.cpp]: I'll change name of the function

File: roboteam_ai/src/control/ballHandling/BallHandlePosControl.cpp (hunk 6 above)
Comment (ThijsLuttikhuis) [reply roboteam_ai/src/control/ballHandling/BallHandlePosControl.cpp]: fixed, renamed

File: roboteam_ai/src/control/ballHandling/BallHandlePosControl.cpp [hunk 11]
```diff
@@ -78,9 +82,17 @@ RobotCommand BallHandlePosControl::getRobotCommand(const RobotPtr &r, const Vect
     // if we do not have the ball yet, go get it
     bool robotDoesNotHaveBall = ! robot->hasBall();
     bool robotIsTooFarFromBall = (robot->pos - ball->pos).length2() > maxBallDistance*maxBallDistance;
+    bool robotIsTouchingBall =
+            (robot->pos - ball->pos).length() < (Constants::ROBOT_RADIUS() + Constants::BALL_RADIUS())*1.05;
```
Comment (rolfvdhulst) [inline roboteam_ai/src/control/ballHandling/BallHandlePosControl.cpp]: magic number

File: roboteam_ai/src/control/ballHandling/BallHandlePosControl.cpp [hunk 12]
```diff
@@ -250,12 +275,12 @@ RobotCommand BallHandlePosControl::goToMovingBall() {
     Vector2 projectionPosition = ballLine.project(robot->pos);
     double robotToProjectionDistance = (projectionPosition - robot->pos).length();
     double ballToProjectionDistance = (projectionPosition - ball->pos).length();
-    const double averageRobotVelocity = 1.5; // ms-1
+    const double averageRobotInterceptVelocity = 1.5; // ms-1
```
Comment (rolfvdhulst) [inline roboteam_ai/src/control/ballHandling/BallHandlePosControl.cpp]: capitalize constants

File: roboteam_ai/src/control/ballHandling/BallHandlePosControl.cpp (hunk 2 above)
Comment (rolfvdhulst) [reply roboteam_ai/src/control/ballHandling/BallHandlePosControl.cpp]: You did not fix this.

File: roboteam_ai/src/world/Ball.cpp [hunk 13]
```diff
@@ -168,30 +170,64 @@ void Ball::updateBallPosition(const Ball &oldBall, const WorldData &worldData) {
...
```
Comment (rolfvdhulst) [inline roboteam_ai/src/world/Ball.cpp]: Can you atleast put a comment with a warning here that is very explicit that this is dangerous/ is changing our world definition?

File: roboteam_ai/src/control/ballHandling/BallHandlePosControl.cpp (hunk 11 above)
Comment (ThijsLuttikhuis) [reply roboteam_ai/src/control/ballHandling/BallHandlePosControl.cpp]: I refactored all of constants.h, but I fixed it.

File: roboteam_ai/src/control/ballHandling/BallHandlePosControl.cpp (hunk 12 above)
Comment (ThijsLuttikhuis) [reply roboteam_ai/src/control/ballHandling/BallHandlePosControl.cpp]: fixed

File: roboteam_ai/src/world/Ball.cpp (hunk 13 above)
Comment (ThijsLuttikhuis) [reply roboteam_ai/src/world/Ball.cpp]: sure

File: roboteam_ai/src/control/ballHandling/BallHandlePosControl.cpp (hunk 1 above)
Comment (ThijsLuttikhuis) [reply roboteam_ai/src/control/ballHandling/BallHandlePosControl.cpp]: the function call is not a static function call, although it might look like it is.
 in fact, you can't even call non static functions in the first place.

File: roboteam_ai/src/control/ballHandling/BallHandlePosControl.cpp (hunk 2 above)
Comment (ThijsLuttikhuis) [reply roboteam_ai/src/control/ballHandling/BallHandlePosControl.cpp]: see the above comments

File: roboteam_ai/test/helpers/WorldHelper.cpp [hunk 14]
```diff
@@ -192,4 +192,17 @@ std::pair<roboteam_msgs::World, int> WorldHelper::getWorldMsgWhereRobotHasBall(i
...
+    roboteam_msgs::World worldMsg;
+    roboteam_msgs::WorldBall ball;
+    ball.pos.x = 9e9;
+    ball.pos.y = 9e9;
+    ball.vel.x = 9e9;
+    ball.vel.y = 9e9;
+    ball.existence = 99999;
+    ball.visible = true;
+    worldMsg.ball = ball;
+    rtt::ai::world::world->updateWorld(worldMsg);
```
Comment (rolfvdhulst) [inline roboteam_ai/test/helpers/WorldHelper.cpp]: This is really ugly. The worldHelper is not build as a controller of the world but rather something that simulates world messages. This also does not properly reset the world as the name implies since there is still history (for example the ball model still gets updated based compared to the previous one. If you want a reset function, I would make it in our AI world, and not do it like this.

File: roboteam_ai/src/world/World.cpp [hunk 15]
```diff
@@ -9,7 +9,7 @@ namespace world {
 World worldObj;
 World* world = &worldObj;
 
-void World::updateWorld(const roboteam_msgs::World &message) {
+void World::updateWorld(const roboteam_msgs::World &message, bool applyBallFilter) {
```
Comment (rolfvdhulst) [inline roboteam_ai/src/world/World.cpp]: Why?

File: roboteam_ai/src/world/Ball.h [hunk 16]
```diff
@@ -27,7 +27,7 @@ class Ball {
         bool dribbledNow = false;
         double lastKickVel = 0;
         int ballStraightTicks = 0;
-        Vector2 ballStillPosition = Vector2();
+        Vector2 ballStillPosition = {9e9, 9e9};
```
Comment (rolfvdhulst) [inline roboteam_ai/src/world/Ball.h]: This is bad, because it means robots will drive out off the field by default.

File: roboteam_ai/src/world/Ball.h [hunk 17]
```diff
@@ -39,7 +39,7 @@ class Ball {
         Ball();
         // Ball(const Ball &copy) = default;
         explicit Ball(const roboteam_msgs::WorldBall &copy);
-        void updateBall(const BallPtr &oldBall, const WorldData &worldData);
+        void updateBall(const BallPtr &oldBall, const WorldData &worldData, bool applyBallFilter = true);
```
Comment (rolfvdhulst) [inline roboteam_ai/src/world/Ball.h]: This is not the way to do it. What if you don't call the filter for some time and then reset it again? We either use it or we do not.

Comment (ThijsLuttikhuis) [reply roboteam_ai/src/world/Ball.h]: the filter is normally always applied, but it breaks some tests which use linear extrapolation of the ball velocity

File: roboteam_ai/src/conditions/IsRobotClosestToBall.cpp [hunk 18]
```diff
@@ -16,16 +17,20 @@ IsRobotClosestToBall::IsRobotClosestToBall(std::string name, bt::Blackboard::Ptr
 : Condition(std::move(name), std::move(blackboard)) { }
 
 bt::Node::Status IsRobotClosestToBall::onUpdate() {
-    Vector2 ballPos(ball->pos);
-
-    if (properties->hasDouble("secondsAhead")) {
+    Vector2 ballPos;
+    if (properties->getBool("atBallStillPosition")) {
+        ballPos = ball->getBallStillPosition();
```
Comment (mrlukasbos) [inline roboteam_ai/src/conditions/IsRobotClosestToBall.cpp]: I am not a fan of blackboards and I think a seperate skill would be easier to read the trees later on

File: roboteam_ai/src/control/ballHandling/BallHandlePosControl.cpp (hunk 2 above)
Comment (mrlukasbos) [reply roboteam_ai/src/control/ballHandling/BallHandlePosControl.cpp]: this is a call to the baseclass and is thus happening inside the robot object ;-) 

Although the name is confusing. A comment would be great.

Comment (mrlukasbos) [reply roboteam_ai/src/control/ballHandling/BallHandlePosControl.cpp]: (the weird notation is because the function is overridden in ballhandle)

File: roboteam_ai/src/utilities/Constants.h [hunk 19]
```diff
@@ -11,62 +11,67 @@ namespace rtt {
...
         static double MAX_ACC_LOWER();  // Maximum acceleration for moving in the sideways direction
         static double MAX_DEC_UPPER();  // Maximum deceleration for moving in the forward direction
         static double MAX_DEC_LOWER();  // Maximum deceleration for moving in the sideways direction
 
-        static double MAX_VEL_BALLPLACEMENT();
-        static double MAX_ANGULAR_VELOCITY();    // Rad per second
-        static double ROBOT_RADIUS(); // TODO: Need to test if world_state agrees with this definition of the centre of the robot
-        static double ROBOT_RADIUS_MAX();
+        /// ROBOT SPECS ///
+        static constexpr double ROBOT_RADIUS()              { return 0.089; };
```
Comment (mrlukasbos) [inline roboteam_ai/src/utilities/Constants.h]: this is bad for compilation time if we wanna change these constants, although i do see that we don't change them.
